src/data/list/indexes.lean

```diff
@@ -0,0 +1,168 @@
+import data.list.basic
+import data.list.defs
+import data.list.range
+import data.list.zip
+import logic.basic
+
+universes u v
+
+open function
+
+namespace list
+
+variables {α : Type u} {β : Type v}
+
+
+section foldr_with_index
+
+/-- Specification of `foldr_with_index_aux`. -/
+def foldr_with_index_aux_spec (f : ℕ → α → β → β) (start : ℕ) (b : β)
+  (as : list α) : β :=
+foldr (uncurry f) b $ zip (range' start as.length) as
+
+theorem foldr_with_index_aux_spec_cons (f : ℕ → α → β → β) (start b a as)
+  : foldr_with_index_aux_spec f start b (a :: as)
+  = f start a (foldr_with_index_aux_spec f (start + 1) b as) := rfl
+
+theorem foldr_with_index_aux_correct (f : ℕ → α → β → β) (start b as)
+  : foldr_with_index_aux f start b as = foldr_with_index_aux_spec f start b as :=
+by induction as generalizing start;
+    [refl, simp only [foldr_with_index_aux, foldr_with_index_aux_spec_cons, *]]
+
+theorem foldr_with_index_correct' (f : ℕ → α → β → β) (b : β) (as : list α)
+  : foldr_with_index f b as = foldr (uncurry f) b (zip (range as.length) as) :=
+by simp only
+    [ foldr_with_index, foldr_with_index_aux_spec, foldr_with_index_aux_correct
+    , ←range_eq_range' ]
+
+end foldr_with_index
+
+
+theorem indexes_correct (as : list α) : indexes as = range as.length :=
+begin
+  suffices h : foldr (cons ∘ prod.fst) nil ((range as.length).zip as) = range as.length,
+  { simp [indexes, foldr_with_index_correct'], exact h },
+  rw ←foldr_map,
+  simp [map_fst_zip]
+end
```

Comment:
Hopefully remove, if `indexes` is removed?

Reply:
Removed.